rts/motoko-rts/src/gc/incremental/partitioned_heap.rs

```diff
@@ -0,0 +1,703 @@
+                partition.free();
+            }
+        }
+        self.evacuating = false;
+        self.bitmap_allocation_pointer = 0;
+        debug_assert!(self.gc_running);
+        self.gc_running = false;
+    }
+
+    pub fn updates_needed(&self) -> bool {
+        self.evacuating
+    }
+
+    fn allocation_partition(&mut self) -> &mut Partition {
+        &mut self.partitions[self.allocation_index]
+    }
+
+    pub fn is_allocation_partition(&self, index: usize) -> bool {
+        self.allocation_index == index
+    }
+
+    unsafe fn allocate_free_partition<M: Memory>(
+        &mut self,
+        mem: &mut M,
+        requested_space: usize,
+    ) -> &mut Partition {
+        let bitmap_address = if self.gc_running {
+            self.allocate_bitmap(mem)
+        } else {
+            null_mut()
+        };
+        for partition in &mut self.partitions {
+            if partition.free && partition.free_size() >= requested_space {
+                debug_assert_eq!(partition.dynamic_size, 0);
+                partition.free = false;
+                if bitmap_address != null_mut() {
+                    partition.bitmap.assign(bitmap_address);
+                }
+                return partition;
+            }
+        }
+        rts_trap_with("Cannot grow memory");
+    }
+
+    pub fn occupied_size(&self) -> Bytes<u32> {
+        let occupied_size: usize = self
+            .partitions
+            .iter()
+            .map(|partition| partition.static_size + partition.dynamic_size)
+            .sum();
+        Bytes(occupied_size as u32)
+    }
+
+    pub fn reclaimed_size(&self) -> Bytes<u64> {
+        Bytes(self.reclaimed)
+    }
+
+    pub unsafe fn allocate<M: Memory>(&mut self, mem: &mut M, words: Words<u32>) -> Value {
+        let size = words.to_bytes().as_usize();
+        if size <= PARTITION_SIZE {
+            self.allocate_normal_object(mem, size)
+        } else {
+            self.allocate_large_object(mem, size)
+        }
+    }
+
+    unsafe fn allocate_normal_object<M: Memory>(&mut self, mem: &mut M, size: usize) -> Value {
+        debug_assert!(size <= PARTITION_SIZE);
+        let mut allocation_partition = self.allocation_partition();
+        debug_assert!(!allocation_partition.free);
+        let heap_pointer = allocation_partition.dynamic_space_end();
+        debug_assert!(size <= allocation_partition.end_address());
+        if heap_pointer <= allocation_partition.end_address() - size {
+            (*allocation_partition).dynamic_size += size;
+            Value::from_ptr(heap_pointer)
+        } else {
+            self.allocate_in_new_partition(mem, size)
+        }
+    }
+
+    // Significant performance gain by not inlining.
+    #[inline(never)]
+    unsafe fn allocate_in_new_partition<M: Memory>(&mut self, mem: &mut M, size: usize) -> Value {
+        #[cfg(feature = "memory_check")]
+        self.allocation_partition().clear_free_remainder();
+
```

Comment:
```suggestion

        debug_assert!(size <= PARTITION_SIZE);
```
?

Reply:
I was unsure, as it is already checked in `allocate_normal_object()` and I only introduced this helper function for `allocate_normal_object` as optimization (for non-inlining the infrequent execution path). Worst case, `allocate_free_partition` would trap with "out of memory" if the size invariant would be violated.